File: src/main/java/javabot/operations/URLTitleOperation.java
package javabot.operations;

import javabot.Message;
import javabot.operations.urlcontent.URLContentAnalyzer;
import javabot.operations.urlcontent.URLFromMessageParser;
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.util.EntityUtils;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import javax.inject.Inject;
import java.io.IOException;
import java.net.URL;
import java.util.List;

import static java.lang.String.format;

public class URLTitleOperation extends BotOperation {
    @Inject
    URLContentAnalyzer analyzer;
    @Inject
    URLFromMessageParser parser;

    static final RequestConfig requestConfig = RequestConfig.custom()
                                                            .setConnectionRequestTimeout(5000)
                                                            .setConnectTimeout(5000)
                                                            .setSocketTimeout(5000)
                                                            .build();

    @Override
    public boolean handleChannelMessage(final Message event) {
        final String message = event.getValue();
        try {
            for (URL url: parser.urlsFromMessage(message)) {
                findTitle(event, url.toString(), true);
                return true;
            }
            return false;
        } catch (IOException ignored) {
            ignored.printStackTrace();
            return false;
        }
    }

    private void findTitle(Message event, String url, boolean loop) throws IOException {
        if (analyzer.precheck(url)) {
            try (CloseableHttpClient client = HttpClientBuilder
                                                  .create()
                                                  .setDefaultRequestConfig(requestConfig)
                                                  .build()) {
                HttpGet httpget = new HttpGet(url);
                httpget.setHeader("User-Agent", "Mozilla/5.0 (Windows NT 6.3; rv:36.0) Gecko/20100101 Firefox/36.0");
                HttpResponse response = client.execute(httpget);
                HttpEntity entity = response.getEntity();
                try {
                    if (!(response.getStatusLine().getStatusCode() == 404 ||
                          response.getStatusLine().getStatusCode() == 403)) {
                        if (entity != null) {
                            Document doc = Jsoup.parse(EntityUtils.toString(entity));
                            String title = clean(doc.title());
                            if (analyzer.check(url, title)) {
                                getBot().postMessageToChannel(event, format("%s'%s title: %s",
                                                                            event.getUser().getNick(),
                                                                            event.getUser().getNick().endsWith("s") ? "" : "s",
                                                                            title));
                            }
                        }
                    }
                } finally {
                    EntityUtils.consume(entity);
                }
            } catch (IOException ioe) {
                if (loop && !url.substring(0, 10).contains("//www.")) {
                    String tUrl = url.replace("//", "//www.");
                    findTitle(event, tUrl, false);
                }
            } catch (IllegalArgumentException ignored) {
            }
        }
    }

    private String clean(String title) {
        StringBuilder sb = new StringBuilder();
        title.chars().filter(i -> i < 127).forEach(i -> sb.append((char) i));
        return sb.toString();
    }
}

File: src/main/kotlin/javabot/operations/URLTitleOperation.kt
package javabot.operations

import java.io.IOException
import java.net.URI
import javabot.Javabot
import javabot.Message
import javabot.dao.AdminDao
import javabot.operations.urlcontent.URLContentAnalyzer
import javabot.operations.urlcontent.URLFromMessageParser
import javabot.service.HttpService
import javabot.service.TwitterService
import javax.inject.Inject
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.nodes.TextNode
import org.jsoup.select.Elements

class URLTitleOperation
@Inject
constructor(
    bot: Javabot,
    adminDao: AdminDao,
    private val analyzer: URLContentAnalyzer,
    private val parser: URLFromMessageParser,
    private val httpService: HttpService,
    private val twitterService: TwitterService,
) : BotOperation(bot, adminDao) {

    override fun handleChannelMessage(event: Message): List<Message> {
        val responses = arrayListOf<Message>()
        val message = event.value
        try {
            val titlesToPost =
                parser
                    .urlsFromMessage(message)
                    .map { it.toString() }
                    .distinct()
                    .mapNotNull { s -> findTitle(s, true) }
                    .filter { s -> s.length >= 20 }
            return if (titlesToPost.isEmpty()) {
                responses
            } else {
                postMessageToChannel(responses, titlesToPost, event)
                responses
            }
        } catch (ignored: Exception) {
            return responses
        }
    }

    private fun postMessageToChannel(
        responses: MutableList<Message>,
        titlesToPost: List<String>,
        event: Message,
    ) {
        val distinctTitles = titlesToPost.distinct()
        val title = if (distinctTitles.size == 1) "title" else "titles"
        responses.add(
            Message(
                event,
                "${event.user.nick}'s $title: " +
                    distinctTitles.joinToString(
                        " | ",
                        transform = { s ->
                            if (s.length > 240) {
                                """"${s.substring(0, 239)}...""""
                            } else {
                                """"$s""""
                            }
                        },
                    ),
            )
        )
    }

    fun hostBasename(host: String, sections: Int = 2): String {
        return host.split(".").takeLast(sections).joinToString(".").lowercase()
    }

    private fun findTitle(url: String, loop: Boolean): String? {
        val twitterUrls = listOf("twitter.com", "x.com")
        if (analyzer.precheck(url)) {
            @Suppress("GrazieInspection")
            try {
                val typedUrl = URI(url).toURL()
                // there used to be parsing here for youtube but they use <title> now
                if (twitterUrls.contains(typedUrl.host.lowercase())) {
                    return findTwitterTitle(url)
                }
                val document = Jsoup.parse(httpService.get(url))
                var title = parseTitle(document)
                title = clean(title)
                return if (analyzer.check(url, title)) title else null
            } catch (ioe: IOException) {
                return if (loop && !url.take(10).contains("//www.")) {
                    val tUrl = url.replace("//", "//www.")
                    findTitle(tUrl, false)
                } else {
                    null
                }
            } catch (ignored: IllegalArgumentException) {
                return null
            }
        } else {
            return null
        }
    }

    private fun findTwitterTitle(url: String): String {
        @Suppress("GrazieInspection")
        if (twitterService.isEnabled()) {
            if (url.contains("status")) {
                val regex = Regex("status/\\d++")
                val match = regex.find(url) // if match isn't null, we have a status number now, woo
                if (match != null) {
                    val id = (match.value.substring("status/".length)).toLong()
                    return twitterService.getStatus(id) ?: ""
                }
            }
            // what happens if it's actually a twitter user? Who cares, we haven't seen that yet in
            // use, right?
        }
        return ""
    }

    private fun fixTwitterLinks(elt: Element) {
        elt.children().forEach { fixTwitterLinks(it) }
        val twitterLinks = elt.getElementsByAttributeValue("data-pre-embedded", "true")
        twitterLinks.forEach { link -> link.insertChildren(0, TextNode(" ")) }
    }

    private fun parseTitle(doc: Document): String {
        val activityPubHeader = doc.select("link[type=\"application/activity+json\"]")
        if (activityPubHeader.isNotEmpty()) {
            val description = doc.head().select("meta[property=\"og:description\"]")
            if (description.isNotEmpty()) {
                val text = description.first()!!.attr("content")
                return text
            }
        }
        val header = doc.select("meta[property=\"og:title\"]")
        return if (header.isNotEmpty()) {
            val body: Elements =
                doc.select(".permalink-tweet-container .js-tweet-text-container") ?: Elements()
            if (body.isNotEmpty()) {
                body.forEach { fixTwitterLinks(it) }
                String.format("%s: \"%s\"", header.first()?.attr("content"), body.first()?.text())
            } else {
                doc.title()
            }
        } else {
            doc.title()
        }
    }

    private fun clean(title: String) = title // IRC can handle real characters
}
